accepted/2022/simplify-rid-model.md

```diff
@@ -0,0 +1,244 @@
+
+- https://peps.python.org/pep-0600/
+- https://peps.python.org/pep-0656/
+- https://doc.rust-lang.org/rustc/platform-support.html
+
+## Biggest problems
+
+The following are the biggest problems we see:
+
+- We get requests to add new RIDs ([Asianux Server 8](https://github.com/dotnet/runtime/issues/2129), [CBL-Mariner](https://github.com/dotnet/runtime/issues/65566), [Anolis Linux](https://github.com/dotnet/runtime/pull/66132)), with no end in sight.
+- Existing RID authoring sometimes [doesn't work with new OS releases](https://github.com/dotnet/runtime/issues/65152).
+- Reasoning about [portable vs non-portable Linux RIDs](https://github.com/dotnet/runtime/pull/62942).
+- We have to read and process `runtime.json` at application startup, which has a (unmeasured) performance impact.
+- The fact that the RID graph is so large (and continuing to grow) demonstrates that we chose a poor design point.
+
+## General approach
+
+- Ensure `runtimes.json` is correct.
+- Freeze `runtimes.json`.
+- Continue to use `runtimes.json` for all NuGet scenarios.
+- Disable using `runtimes.json` for all host scenarios, by default (starting with a TBD .NET version).
+- Enable a host compatibility mode that uses `runtimes.json` via MSBuild property (which writes to), `runtimeconfig.json`, and a host CLI argument.
+- Implement a new algorithmic host scheme for RID selection, enabled by default.
+
+The new scheme will be based on the following:
+
+- Each host is built with a hard-coded set of RIDs that it probes for, both triplets and singles.
+- Each host is already built for a specific RID triple, so this change is an evolutionary change.
+- There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) that works the same x64, Arm64 and x86. The same is true for macOS and Linux.
+- The RID that `dotnet --info` returns will always match the first in this hard-coded set of RIDs.
+
+Let's assume that an app is running on Alpine 3.16, the host will probe for the following RIDs:
+
+- `linux-musl-x64`
+- `unix`
+
+Ubuntu would be similar:
+
+- `linux-arm64`
+- `unix`
+
+macOS would be similar:
+
+- `osx-arm64`
+- `unix`
+
+Windows would similar:
+
+- `win-x86`
+- `win`
+
+Note: `win7` and `win10` also exist (as processor-specific and -agnostic). Ideally, we don't have to support those in the host-specific scheme, but we need to do some research on that.
+
+Note: A mix of processor-specific RIDs are used above, purely to demonstrate the range of processor-types supported. Each operating system supports a variety of processor-types, and this is codified in their associated RIDs.
+
+More generally, the host will probe for:
+
+- This environment, RID triplet (OS-CRuntime-arch)
+- This environment, RID single (OS-only).
+
+Note: The abstract `unix` RID is used for macOS and Linux, instead of a concrete RID for those OSes.
```

Comment:
What if I wanted to create a managed .dll that worked across all `macOS` architectures? The same scenario as the above:

> There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) 

Except for `macOS`-specific. Like P/Invoking into macOS native APIs.

The same scenario can be made for Linux and reading certain files like `/proc/self`, etc.

Reply:
Or is the expectation that .NET Libraries should be using TFMs (`net6.0-macos`, `net6.0-windows`, etc) for those scenarios?

Reply:
My understanding is that the use of macOS-specific code is limited and that we rely primarily on `unix` in our own codebase. As a result, I saw `osx` as something we should wait for more concrete scenarios. If that's wrong, we can add `osx` to the macOS probing list now.

The TFMs are really fully separate. They can be use for the purpose you intend, but not for code that is intended as `net6.0`. For that, you need to use RIDs, as you know.

Reply:
I'll add this to the spec as something to resolve. It will mostly go away, since `osx-x64` will eventually age out.